drivers/sensor/tsl2591/tsl2591.c

```diff
@@ -0,0 +1,535 @@
+/*
+ * Copyright (c) 2023 Kurtis Dinelle
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ams_tsl2591
+
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include "tsl2591.h"
+
+LOG_MODULE_REGISTER(TSL2591, CONFIG_SENSOR_LOG_LEVEL);
+
+static int tsl2591_reg_read(const struct device *dev, uint8_t reg, uint8_t *buf, uint8_t size)
+{
+	const struct tsl2591_config *config = dev->config;
+	uint8_t cmd = TSL2591_NORMAL_CMD | reg;
+
+	return i2c_write_read_dt(&config->i2c, &cmd, 1U, buf, size);
+}
+
+static int tsl2591_reg_write(const struct device *dev, uint8_t reg, uint8_t val)
+{
+	const struct tsl2591_config *config = dev->config;
+	uint8_t cmd[2] = {TSL2591_NORMAL_CMD | reg, val};
+
+	return i2c_write_dt(&config->i2c, cmd, 2U);
+}
+
+int tsl2591_reg_update(const struct device *dev, uint8_t reg, uint8_t mask, uint8_t val)
+{
+	uint8_t old_value, new_value;
+	int ret;
+
+	ret = tsl2591_reg_read(dev, reg, &old_value, 1U);
+	if (ret < 0) {
+		return ret;
+	}
+
+	new_value = (old_value & ~mask) | (val & mask);
+	if (new_value == old_value) {
+		return 0;
+	}
+
+	return tsl2591_reg_write(dev, reg, new_value);
+}
+
+static int tsl2591_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct tsl2591_data *data = dev->data;
+	uint8_t als_data[4];
+	int ret;
+
+#ifdef CONFIG_TSL2591_FETCH_WAIT
+	uint8_t status;
+
+	ret = tsl2591_reg_read(dev, TSL2591_REG_STATUS, &status, 1U);
+	if (ret < 0) {
+		LOG_ERR("Failed to read status register");
+		return ret;
+	}
+
+	/* Check if ALS has completed an integration cycle since AEN asserted.
+	 * If not, sleep for the duration of an integration cycle to ensure valid reading.
+	 */
+	if ((status & TSL2591_AVALID_MASK) != 1) {
```

Comment:
would probably be better to not hard code the 1 here, no?

Reply:
Yep, good point, thank you for the review! The AVALID bitmask is just a single bit, so I replaced this line with:

```c
if (!(status & TSL2591_AVALID_MASK)) {
```

Since we don't care about a particular value, only if the bit is set or not. Please let me know if a different approach is preferred!

Reply:
great, thank you!